### PR TITLE
Add extra parameters to FindCommandParser to parse

### LINKED ISSUE #104: Add FilterCommand

### PR DESCRIPTION
**Pull request by jevan13**
## Describe your changes
- Add extra parameters for FindCommandParser.parse()
- Parse using CLI Syntax
- Create separate ContainsKeywordsPredicate classes for simplicity
- Updated Testcases

## Issue ticket number and link
closes #104 

## Checklist before requesting a review
- [X] I have performed a self-review of my code.
- [X] If it is a core feature, I have added thorough tests.

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/FindCommand.java
Comment (InfinityTwo) [inline src/main/java/seedu/address/logic/commands/FindCommand.java:30]: Perhaps the name declaration here could be changed to SCREAMING_SNAKE_CASE?

File: src/main/java/seedu/address/logic/parser/FindCommandParser.java
Comment (InfinityTwo) [inline src/main/java/seedu/address/logic/parser/FindCommandParser.java]: Perhaps lines 56-90 could be abstracted into a function that dynamically handles the differences? The statements inside every if statement look the same and perhaps the methods for predicateContainer could be abstracted into one function inside the class to handle the input types?

File: src/main/java/seedu/address/model/person/DepartmentContainsKeywordsPredicate.java
Comment (InfinityTwo) [inline src/main/java/seedu/address/model/person/DepartmentContainsKeywordsPredicate.java:13]: Ditto on naming constants

File: src/main/java/seedu/address/model/person/EmailContainsKeywordsPredicate.java
Comment (InfinityTwo) [inline src/main/java/seedu/address/model/person/EmailContainsKeywordsPredicate.java:13]: Ditto

File: src/main/java/seedu/address/model/person/PhoneContainsKeywordsPredicate.java
Comment (InfinityTwo) [inline src/main/java/seedu/address/model/person/PhoneContainsKeywordsPredicate.java:13]: Ditto

File: src/main/java/seedu/address/model/person/PredicateContainer.java
Comment (InfinityTwo) [inline src/main/java/seedu/address/model/person/PredicateContainer.java]: Perhaps there should be an explanation/JavaDoc why the constructor is empty?

File: src/main/java/seedu/address/model/person/RoleContainsKeywordsPredicate.java
Comment (InfinityTwo) [inline src/main/java/seedu/address/model/person/RoleContainsKeywordsPredicate.java:13]: Ditto

File: src/test/java/seedu/address/logic/parser/AddressBookParserTest.java
Comment (InfinityTwo) [inline src/test/java/seedu/address/logic/parser/AddressBookParserTest.java:120]: I like the comprehensive testing for each Find class but perhaps instead of appending 1, 2, 3 or 4 to the names of the variable, there could be a better naming convention?

File: src/test/java/seedu/address/logic/parser/FindCommandParserTest.java
Comment (InfinityTwo) [inline src/test/java/seedu/address/logic/parser/FindCommandParserTest.java:39]: Any reason why you didn't import the valid person's name and/or substring the name to avoid regressions?

Comment (InfinityTwo) [reply src/test/java/seedu/address/logic/parser/FindCommandParserTest.java:39]: The same question goes for their details like address, contact, email, etc.

File: src/test/java/seedu/address/logic/parser/FindCommandParserTest.java
Comment (InfinityTwo) [inline src/test/java/seedu/address/logic/parser/FindCommandParserTest.java:109]: Perhaps you can also import the constant for all, ph and as well?

File: src/test/java/seedu/address/model/person/DepartmentContainsKeywordsPredicateTest.java
Comment (InfinityTwo) [inline src/test/java/seedu/address/model/person/DepartmentContainsKeywordsPredicateTest.java:39]: I like the comprehensive test cases but perhaps you could use more more edge case test cases like null and empty strings here for all the other test cases?

File: src/test/java/seedu/address/model/person/NameContainsKeywordsPredicateTest.java
Comment (InfinityTwo) [inline src/test/java/seedu/address/model/person/NameContainsKeywordsPredicateTest.java:46]: Ditto for imports

**Review by InfinityTwo — CHANGES_REQUESTED** [review]
I think your PR generally looks comprehensive but could use less hardcoding of existing constants/test persons. I think the PR is good to go once that, merge conflicts and codecov are fixed.

File: src/main/java/seedu/address/logic/parser/FindCommandParser.java
Comment (jevan13) [reply src/main/java/seedu/address/logic/parser/FindCommandParser.java]: Updated. FindCommandParser will use a PredicateContainer static method to extract the values out from the argumentMultimap. Add more methods in PredicateContainer to set the Predicate fields using the argumentMultimap.

File: src/main/java/seedu/address/model/person/PredicateContainer.java
Comment (jevan13) [reply src/main/java/seedu/address/model/person/PredicateContainer.java]: Will remove constructor

File: src/test/java/seedu/address/logic/parser/FindCommandParserTest.java
Comment (jevan13) [reply src/test/java/seedu/address/logic/parser/FindCommandParserTest.java:39]: I feel its better for readability, to see the actual inputs, actual and expected results.

File: src/test/java/seedu/address/model/person/DepartmentContainsKeywordsPredicateTest.java
Comment (jevan13) [reply src/test/java/seedu/address/model/person/DepartmentContainsKeywordsPredicateTest.java:39]: What other test cases?

Comment (InfinityTwo) [reply src/test/java/seedu/address/model/person/DepartmentContainsKeywordsPredicateTest.java:39]: Every other test case without null tests

**Review by InfinityTwo — APPROVED** [review]

**Pull request merged**
